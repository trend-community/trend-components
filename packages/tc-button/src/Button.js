import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

const classNames = {
  BASE: 'tc-Button',
  ICON: 'tc-Button-icon',
  ACCENT: 'tc-Button--accent',
  COMPACT: 'tc-Button--compact',
  CTAB: 'tc-Button--ctab',
  FLAT: 'tc-Button--flat',
  GHOST: 'tc-Button--ghost'
};

const modifiers = {
  ACCENT: 'accent',
  CTAB: 'ctab',
  FLAT: 'flat',
  GHOST: 'ghost',
  COMPACT: 'compact'
};

class Button extends PureComponent {
  static propTypes = {
    accent: PropTypes.bool,
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.func
    ]).isRequired,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    type: PropTypes.oneOf(['button', 'submit', 'reset']),
    size: PropTypes.oneOf([modifiers.COMPACT]),
    variant: PropTypes.oneOf([modifiers.CTAB, modifiers.FLAT, modifiers.GHOST])
  }

  static defaultProps = {
    accent: false,
    type: 'button'
  }

  getButtonIconProps(props = {}) {
    return {
      ...props,
      className: cn(classNames.ICON, props.className),
      'aria-hidden': true
    };
  }

  api() {
    return {
      getButtonIconProps: this.getButtonIconProps
    };
  }

  render() {
    const {
      accent,
      children,
      className,
      size,
      type,
      variant,
      ...rest
    } = this.props;

    return <button
      {...rest}
      className={cn(
        classNames.BASE,
        className, {
          [classNames.ACCENT]: accent,
          [classNames.COMPACT]: size === modifiers.COMPACT,
          [classNames.CTAB]: variant === modifiers.CTAB,
          [classNames.FLAT]: variant === modifiers.FLAT,
          [classNames.GHOST]: variant === modifiers.GHOST,
        }
      )}
      type={type}>
      {typeof children === 'function'
        ? children(this.api())
        : children
      }
    </button>;
  }
}

export default Button;
