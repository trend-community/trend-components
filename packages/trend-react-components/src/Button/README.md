# React Buttons

Create **TREND** buttons using react.

## Installation

```bash
npm install @trend/react-button
```

## Basic Usage

With a module bunlder like [webpack](https://webpack.js.org/), use as you would anything else:

```javascript
// Using ES6 modules.
@import React from 'react';
@import Button from 'trend-react-components/Button';

// Render a basic TREND component button.
function MyButton(props) {
  return <Button {...props}>My Button</Button>;
}

// Will output:
// <button class="tc-Button" type="button">My Button</button>

// Render a ctab button using modifiers as a string.
functon CtabButton() {
  return <Button modifiers='ctab'>Ctab Button</Button>;
}

// Will output:
// <button class="tc-Button tc-Button--ctab" type="button">My Button</button>

// Render a flat accent button using modifiers as a array.
functon CtabButton() {
  return <Button modifiers={['flat', 'accent']}>Ctab Button</Button>;
}

// Will output:
// <button class="tc-Button tc-Button--flat tc-Button--accent" type="button">
//   My Button
// </button>

// Render children as a function.
function ButtonWithIcon() {
  return <Button>
    {({ getButtonIconProps }) =>
      <React.Fragment>
        <svg {...getButtonIconProps()}>//...<svg>
        Button with icon
      </React.Fragment>
    }
  </Button>
}

// Will output:
// <button class="tc-Button" type="button">
//   <svg class="tc-Button-icon" aria-hidden="true" ...>//...</svg>
//   Button with icon
// </button>
```

## Props:

### className

> `string` | optional, no default

Will automatically add `tc-Button` to rendered button element.

### disabled

> `boolean` | defaults to `false`

### modifiers

> `string` or `array` | optional

Can be any of "accent", "ctab", "flat", "ghost", and/or "compact."  Will prepend `tc-Button--` to each name passed in.

### type

> `string` | defaults `button`

Can be any one of `button`, `submit`, or `reset`.

### children

> `function({}) or node` | required

Children as function gets called with an object and is useful for rendering a child icon in a button element.


Property | Category | Type | Description
--- | --- | --- | ---
`getButtonIconProps` | prop getter | `function(props: object)` | Will return the props to apply to a icon as a child of a button element.
