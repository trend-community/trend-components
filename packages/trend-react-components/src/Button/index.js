import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

const classNames = {
  BASE: 'tc-Button',
  ICON: 'tc-Button-icon'
};
const MODIFIERS = ['accent', 'ctab', 'flat', 'ghost', 'compact'];

function isModifier(modifier) {
  return MODIFIERS.indexOf(modifier) !== -1;
}

function formatModifier(modifier) {
  return `${classNames.BASE}--${modifier}`;
}

function buildModifiers(modifiers) {
  const mods = Array.isArray(modifiers) ? modifiers : modifiers.split(' ');

  return mods
    .filter(isModifier)
    .map(formatModifier)
    .join(' ');
}

class Button extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.func
    ]).isRequired,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    modifiers: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string)
    ]),
    type: PropTypes.oneOf(['button', 'submit', 'reset'])
  }

  static defaultProps = {
    modifiers: '',
    type: 'button'
  }

  getButtonIconProps(props = {}) {
    return {
      ...props,
      className: cn(classNames.ICON, props.className),
      'aria-hidden': true
    };
  }

  getButtonHelpers() {
    return {
      getButtonIconProps: this.getButtonIconProps
    };
  }

  render() {
    return this._createComponent(this.props);
  }

  _createComponent(props) {
    const {
      children,
      className,
      modifiers,
      type,
      ...rest
    } = props;

    return <button
      {...rest}
      className={cn(classNames.BASE, buildModifiers(modifiers), className)}
      type={type}>
      {typeof children === 'function'
        ? children(this.getButtonHelpers())
        : children
      }
    </button>;
  }
}

export default Button;
