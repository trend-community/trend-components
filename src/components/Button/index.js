import 'packages/tc-button/index.scss';
export { default } from 'packages/tc-react-button';
