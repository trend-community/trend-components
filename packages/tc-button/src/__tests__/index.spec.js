import React from 'react';
import { shallow } from 'enzyme';

import Button from '../';

describe('tc-button', () => {
  it('should render a button element.', () => {
    const wrapper = shallow(<Button>button</Button>);
    expect(wrapper.type()).toEqual('button');
  });

  it('should have correct default className.', () => {
    const wrapper = shallow(<Button>button</Button>);
    expect(wrapper.prop('className')).toMatchSnapshot();
  });

  it('should set a default "type" prop.', () => {
    const wrapper = shallow(<Button>button</Button>);
    expect(wrapper.prop('type')).toEqual('button');
  });

  it('should disable the button when `disabled`.', () => {
    const wrapper = shallow(<Button disabled>button</Button>);
    expect(wrapper.prop('disabled')).toBeTruthy();
  });

  it('should render a flat accent button.', () => {
    const wrapper = shallow(<Button variant="flat" accent>button</Button>);
    expect(wrapper.prop('className')).toMatchSnapshot();
  });

  it('should render a compact ghost button.', () => {
    const wrapper = shallow(<Button variant="ghost" size="compact" />);
    expect(wrapper.prop('className')).toMatchSnapshot();
  });

  it('should render a disabled ctab accent button.', () => {
    const wrapper = shallow(<Button variant="ctab" accent disabled />);
    expect(wrapper.props()).toMatchSnapshot();
  });

  it('should pass through additional props to button.', () => {
    const wrapper = shallow(<Button id="test" aria-label="test-button" />);
    expect(wrapper.props()).toMatchSnapshot();
  });

  it('should render children.', () => {
    const testCase = 'test';
    const wrapper = shallow(<Button children={testCase} />);
    expect(wrapper.text()).toEqual(testCase);
  });

  it('should expose a static prop getter for icons.', () => {
    expect(Button.getIconProps).toBeTruthy();
    expect(Button.getIconProps()).toMatchSnapshot();
    expect(Button.getIconProps({ id: 'test' })).toMatchSnapshot();
  });
});
