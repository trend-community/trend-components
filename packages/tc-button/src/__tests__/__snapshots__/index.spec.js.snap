// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tc-button should expose a static prop getter for icons. 1`] = `
Object {
  "aria-hidden": true,
  "className": "tc-Button-icon",
}
`;

exports[`tc-button should expose a static prop getter for icons. 2`] = `
Object {
  "aria-hidden": true,
  "className": "tc-Button-icon",
  "id": "test",
}
`;

exports[`tc-button should have correct default className. 1`] = `"tc-Button"`;

exports[`tc-button should pass through additional props to button. 1`] = `
Object {
  "aria-label": "test-button",
  "className": "tc-Button",
  "id": "test",
  "type": "button",
}
`;

exports[`tc-button should render a compact ghost button. 1`] = `"tc-Button tc-Button--compact tc-Button--ghost"`;

exports[`tc-button should render a disabled ctab accent button. 1`] = `
Object {
  "className": "tc-Button tc-Button--accent tc-Button--ctab",
  "disabled": true,
  "type": "button",
}
`;

exports[`tc-button should render a flat accent button. 1`] = `"tc-Button tc-Button--accent tc-Button--flat"`;
